### ISSUE TITLE
BUG: agent not inventory network interfaces without assigned IP address

### ISSUE THREAD

**Issue by mariaczi** [OCSInventory-NG/UnixAgent#207]
### General informations
Operating system : Linux Slackware 
Perl version : 5.26

### OCS Inventory informations
Unix agent version : 2.3.1

### Problem's description
OCS Agent don't collect and send information to OCS Server about the network interfaces which ~~are not UP~~ haven't assigned IP(s).
I will be happy to know, that I have ~~other free (~~ not used ~~)~~ network interfaces on server(s).

**Comment by fbomj** [comment]
Have you errors in ocsinventory-agent.log file?

**Comment by mariaczi** [comment]
No any errors in agent log file.
From ocsagent.log file:
```
    <CONTROLLERS>
      <DRIVER>igb</DRIVER>
      <MANUFACTURER>Intel Corporation I210 Gigabit Network Connection</MANUFACTURER>
      <NAME>Ethernet controller [0200]</NAME>
      <PCIID>8086:1533</PCIID>
      <PCISLOT>02:00.0</PCISLOT>
      <TYPE>rev 03</TYPE>
    </CONTROLLERS>
    <CONTROLLERS>
      <DRIVER>igb</DRIVER>
      <MANUFACTURER>Intel Corporation I210 Gigabit Network Connection</MANUFACTURER>
      <NAME>Ethernet controller [0200]</NAME>
      <PCIID>8086:1533</PCIID>
      <PCISLOT>03:00.0</PCISLOT>
      <TYPE>rev 03</TYPE>
    </CONTROLLERS>
```
And a little below (from the same log file):
```
    <NETWORKS>
      <DESCRIPTION>eth0</DESCRIPTION>
      <DRIVER>igb</DRIVER>
      <DUPLEX>Full</DUPLEX>
      <IPADDRESS>62.aaa.xx.yy</IPADDRESS>
      <IPGATEWAY>62.aaa.xx.z</IPGATEWAY>
      <IPMASK>255.255.255.0</IPMASK>
      <IPSUBNET>62.aaa.xx.0</IPSUBNET>
      <MACADDR>Aa:bb:cc:dd:ee:E8</MACADDR>
      <MTU>1500</MTU>
      <PCISLOT>0000:02:00.0</PCISLOT>
      <SPEED>1 Gbps</SPEED>
      <STATUS>UP</STATUS>
      <TYPE>Ethernet</TYPE>
    </NETWORKS>
    <NETWORKS>
      <DESCRIPTION>eth0</DESCRIPTION>
      <DRIVER>igb</DRIVER>
      <DUPLEX>Full</DUPLEX>
      <IPADDRESS>fe80::aebb:ccff:fedd:eee8</IPADDRESS>
      <IPGATEWAY>::</IPGATEWAY>
      <IPMASK>ffff:ffff:ffff:ffff:0000:0000:0000:0000</IPMASK>
      <IPSUBNET>fe80::</IPSUBNET>
      <MACADDR>Aa:bb:cc:dd:ee:E8</MACADDR>
      <MTU>1500</MTU>
      <PCISLOT>0000:02:00.0</PCISLOT>
      <SPEED>1 Gbps</SPEED>
      <STATUS>UP</STATUS>
      <TYPE>Ethernet</TYPE>
    </NETWORKS>
```
```
# ip link show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether Aa:bb:cc:dd:ee:E8 brd ff:ff:ff:ff:ff:ff
3: eth1: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether Aa:bb:cc:dd:ee:e9 brd ff:ff:ff:ff:ff:ff
```
From OCS server: in attached screen-shot.
![Selection_089](https://user-images.githubusercontent.com/25037132/60420639-7d650080-9be8-11e9-9045-7b9610dfe379.png)

I'll upgrade OCS agent to the latest version and will back with the info about the result (from latest ocs agent version).

**Comment by mariaczi** [comment]
I'm not able to put info about results at the moment because I "catches" the new issue which was reported by the other user (I have added my comment there): https://github.com/OCSInventory-NG/UnixAgent/issues/205#issuecomment-507224187

**Comment by fbomj** [comment]
ipv6 is supported with 2.6.0.

**Comment by mariaczi** [comment]
Hmm, it is a little strange. As you can see (in my first comment), I have details from network interface about ipv6 using OCS Agent version 2.3.1 :) This details were gone after upgrading OCS Agent to version 2.4.2 which inventoried down interfaces :)

**Comment by mariaczi** [comment]
OK, i'm fixed the issue with "uninitialized value" which I had: https://github.com/OCSInventory-NG/UnixAgent/issues/205#issuecomment-508675738
OCS Agent 2.6.0 connect successfully with OCS Server but not inventoried network interfaces which are not UP and don't inventory any information about ipv6 for UP interfaces.
![Selection_095](https://user-images.githubusercontent.com/25037132/60708930-7d7a3e80-9f0f-11e9-982a-b261d5e628da.png)
No any errors in OCS Agent log file. Some detail from agent log file:
```
    <CONTROLLERS>
      <DRIVER>igb</DRIVER>
      <MANUFACTURER>Intel Corporation I210 Gigabit Network Connection</MANUFACTURER>
      <NAME>Ethernet controller [0200]</NAME>
      <PCIID>8086:1533</PCIID>
      <PCISLOT>02:00.0</PCISLOT>
      <TYPE>rev 03</TYPE>
    </CONTROLLERS>
    <CONTROLLERS>
      <DRIVER>igb</DRIVER>
      <MANUFACTURER>Intel Corporation I210 Gigabit Network Connection</MANUFACTURER>
      <NAME>Ethernet controller [0200]</NAME>
      <PCIID>8086:1533</PCIID>
      <PCISLOT>03:00.0</PCISLOT>
      <TYPE>rev 03</TYPE>
    </CONTROLLERS>

...

    </MEMORIES>
    <NETWORKS>
      <DESCRIPTION>eth0</DESCRIPTION>
      <DRIVER>igb</DRIVER>
      <DUPLEX>Full</DUPLEX>
      <IPADDRESS>62.aaa.xx.yy</IPADDRESS>
      <IPGATEWAY>62.aaa.xx.z</IPGATEWAY>
      <IPMASK>255.255.255.0</IPMASK>
      <IPSUBNET>62.aaa.xx.0</IPSUBNET>
      <MACADDR>Aa:bb:cc:dd:ee:e8</MACADDR>
      <MTU>1500</MTU>
      <PCISLOT>0000:02:00.0</PCISLOT>
      <SPEED>1 Gbps</SPEED>
      <STATUS>Up</STATUS>
      <TYPE>ethernet</TYPE>
    </NETWORKS>
    <PORTS>
```

**Comment by mariaczi** [comment]
Fix for OCS Agent v2.6.0 - inventory DOWN network interfaces and ipv6 details
```
diff -u Ocsinventory-Unix-Agent-2.6.0/lib/Ocsinventory/Agent/Backend/OS/Linux/Network/Networks.pm.orig Ocsinventory-Unix-Agent-2.6.0/lib/Ocsinventory/Agent/Backend/OS/Linux/Network/Networks.pm
--- Ocsinventory-Unix-Agent-2.6.0/lib/Ocsinventory/Agent/Backend/OS/Linux/Network/Networks.pm.orig	2019-07-05 15:47:40.347630992 +0200
+++ Ocsinventory-Unix-Agent-2.6.0/lib/Ocsinventory/Agent/Backend/OS/Linux/Network/Networks.pm	2019-07-05 16:28:25.814776607 +0200
@@ -212,7 +212,8 @@
                             MTU => $mtu,
                         });
                     }
-                } elsif ($description && $ipaddress6) {
+                }
+                if ($description && $ipaddress6) {
                     if ($type eq "Wifi") {
                           $common->addNetwork({
                               DESCRIPTION => $description,
@@ -252,6 +253,37 @@
                         });
                     }
                 }
+                if ($description && !$ipaddress && !$ipaddress6) {
+                    if ($type eq "Wifi") {
+                          $common->addNetwork({
+                              DESCRIPTION => $description,
+                              DRIVER => $driver,
+                              MACADDR => $macaddr,
+                              PCISLOT => $pcislot,
+                              STATUS => $status?"Up":"Down",
+                              TYPE => $type,
+                              SPEED => $bitrate,
+                              SSID => $ssid,
+                              BSSID => $bssid,
+                              IEEE => $version,
+                              MODE => $mode,
+                        });
+                    } else {
+                        $common->addNetwork({
+                            DESCRIPTION => $description,
+                            DRIVER => $driver,
+                            MACADDR => $macaddr,
+                            PCISLOT => $pcislot,
+                            STATUS => $status?"Up":"Down",
+                            TYPE => $type,
+                            VIRTUALDEV => $virtualdev,
+                            DUPLEX => $duplex?"Full":"Half",
+                            SPEED => $speed,
+                            MTU => $mtu,
+                        });
+                    }
+                }
+
 
                 # Virtual devices
                 # Reliable way to get the info
@@ -359,7 +391,7 @@
             } elsif ($line =~ /\s+link\/(\S+)/){
                 $type=$1;
                 $macaddr=getMAC($description);
-            } elsif ($line =~ /inet6 (\S+)\/(d{1,2})/i){
+            } elsif ($line =~ /inet6 (\S+)\/(\d{1,2})/i){
                 $ipaddress6=$1;
                 $ipmask6=getIPNetmaskV6($2);
                 $ipsubnet6=getSubnetAddressIPv6($ipaddress6,$ipmask6);
@@ -678,8 +710,8 @@
 
     return undef unless $address && $mask;
 
-    my $binaddress = ip_iptobin($address, 6);
-    my $binmask    = ip_iptobin($mask, 6);
+    my $binaddress = ip_iptobin(ip_expand_address($address,6),6);
+    my $binmask    = ip_iptobin(ip_expand_address($mask,6),6);
     my $binsubnet  = $binaddress & $binmask;
 
     return ip_compress_address(ip_bintoip($binsubnet, 6),6);
```

**Comment by fbomj** [comment]
Hi Mariaczi 

ip informations must be set to 0

**Comment by mariaczi** [comment]
Where? When? If interface is down, it is clear form me, that we haven't setup any IP address on this interface.

**Comment by fbomj** [comment]
you also can have the possibility that you setup an ip address on an interface and have decide to shutdown it. Agree with that?

**Comment by mariaczi** [comment]
Yes, of course.

**Comment by mariaczi** [comment]
The real reason why the agent did not collect and send information about some interfaces is not assigned IP address for them. In that way we shouldn't send fake info abut IP, we must sent the information about the real situation on the server. So, my above attached patch is OK.

**Comment by fbomj** [comment]
Fixed with PR #215
